data/transactions/logic/opcodes.go

```diff
@@ -240,31 +286,55 @@ func imm(name string, kind immKind) immediate {
 	return immediate{name, kind, nil, nil}
 }
 
+type typedList struct {
+	Types   StackTypes // what gets popped from the stack
+	Effects string
+}
+
+// Proto describes the "stack behavior" of an opcode, what it pops as arguments
+// and pushes onto the stack as return values.
+type Proto struct {
+	Arg    typedList // what gets popped from the stack
+	Return typedList // what gets pushed to the stack
+}
+
+func proto(signature string, effects ...string) Proto {
+	parts := strings.Split(signature, ":")
+	if len(parts) != 2 {
+		panic(signature)
+	}
+	var argEffect, retEffect string
+	switch len(effects) {
+	case 1:
+		retEffect = effects[0]
+	case 2:
+		argEffect = effects[0]
+		retEffect = effects[1]
+	}
+	return Proto{
+		Arg:    typedList{parseStackTypes(parts[0]), argEffect},
+		Return: typedList{parseStackTypes(parts[1]), retEffect},
+	}
+}
+
 // OpSpec defines an opcode
 type OpSpec struct {
-	Opcode  byte
-	Name    string
-	op      evalFunc   // evaluate the op
-	asm     asmFunc    // assemble the op
-	dis     disFunc    // disassemble the op
-	Args    StackTypes // what gets popped from the stack
-	Returns StackTypes // what gets pushed to the stack
-	Version uint64     // TEAL version opcode introduced
-	Modes   runMode    // if non-zero, then (mode & Modes) != 0 to allow
-	Details opDetails  // Special cost or bytecode layout considerations
-}
-
-var oneBytes = StackTypes{StackBytes}
-var twoBytes = StackTypes{StackBytes, StackBytes}
-var threeBytes = StackTypes{StackBytes, StackBytes, StackBytes}
-var byteInt = StackTypes{StackBytes, StackUint64}
-var byteIntInt = StackTypes{StackBytes, StackUint64, StackUint64}
-var oneInt = StackTypes{StackUint64}
-var twoInts = StackTypes{StackUint64, StackUint64}
-var oneAny = StackTypes{StackAny}
-var twoAny = StackTypes{StackAny, StackAny}
-var anyInt = StackTypes{StackAny, StackUint64}
-var anyIntInt = StackTypes{StackAny, StackUint64, StackUint64}
+	Opcode byte
+	Name   string
+	op     evalFunc // evaluate the op
+	Proto
+	Version   uint64 // TEAL version opcode introduced
+	OpDetails        // Special cost or bytecode layout considerations
+}
+
+// AlwaysExits is true iff the opcode always ends the program.
+func (spec *OpSpec) AlwaysExits() bool {
```

Comment:
@jannotti For my understanding:  Did benchmarking, concern over struct size or another rationale lead towards a pointer receiver here?

I'm asking for my awareness because my default thought process is to eliminate chance of unintended mutation via value receiver.

Reply:
Object size.  OpSpecs are pretty big, once you consider the size of Proto and OpDetails, which are embedded. It's still unclear to me whether Go can optimize the copy away when inlining a function like this.